File: 0169_Majority_Element/majority.go
package  main

import "fmt"
import "sort"

// 使用排序
func majorityElement2(nums []int) int {
    length := len(nums)
    sort.Ints(nums)
    return nums[length/2]
}

func majorityElement(nums []int) int {
	count := 1
	majority := nums[0]
	for i := 1; i < len(nums); i++ {
		if count==0 {
			majority = nums[i]
		}
		if majority == nums[i] {
			count++
		}else{
			count--
		}
	}

	return majority
}

func main()  {
	var nums = []int{2,2,1,1,1,1,1,2,2}
	var res = majorityElement2(nums)
	fmt.Println(res)
}

File: 0169_Majority_Element/majority.js
/*
 * @Author: Aschenmaker 
 * @Date: 2020-04-21 21:41:39 
 * @Last Modified by: Aschenmaker
 * @Last Modified time: 2020-04-21 21:48:41
 */

// 通过排序算法
// var majorityElement = function(nums) {
// 	nums = nums.sort((a, b) => a - b);
// 	let mid = Math.floor(nums.length / 2);
// 	return nums[mid];
// };

// 投票算法
majorityElement = (nums) => {
	let count = 1;
	let majority = nums[0];
	// 从第二个元素开始投票，相同则++,不同则--；
	for (let i = 1; i < nums.length; i++) {
		if (count == 0) {
			majority = nums[i];
		}
		if (nums[i] == majority) {
			count++;
		} else {
			count--;
		}
		console.log(count);
	}
	return majority;
};

var nums = [ 6, 5, 5 ];

console.log(majorityElement(nums));
